File: frontend/src/lib/queries/validator.ts
// Phase C validator queries — run history + findings.
import 'server-only'
import sql from '@/lib/db'

export type ValidatorRun = {
  id: string
  started_at: Date
  completed_at: Date | null
  status: string
  scope: string
  n_findings: number | null
}

export type ValidatorFinding = {
  id: string
  run_id: string
  finding_class: string
  severity: string
  route: string | null
  surface: string
  identifier: string
  expected_value: string
  actual_value: string
  delta_pct: string | null
  first_seen: Date
  last_seen: Date
}

// One row per (route, surface, severity) — aggregated count + representative sample.
export type FindingGroup = {
  route: string | null
  surface: string
  severity: string
  count: number
  sample_identifier: string | null
  sample_expected: string | null
  sample_actual: string | null
  last_seen: Date
}

export type RouteSummary = {
  route: string
  p0: number
  p1: number
  p2: number
  total: number
}

export type DayTrend = {
  run_date: string
  p0: number
  p1: number
  p2: number
  total: number
}

export async function getRecentValidatorRuns(limit = 20): Promise<ValidatorRun[]> {
  return sql<ValidatorRun[]>`
    SELECT
      id::text            AS id,
      started_at,
      completed_at,
      status,
      scope,
      n_findings
    FROM atlas.atlas_validator_runs
    ORDER BY started_at DESC
    LIMIT ${limit}
  `
}

// All findings from the single most-recent successful frontend_diff run, grouped by (route, surface, severity).
export async function getLatestFrontendFindingGroups(): Promise<FindingGroup[]> {
  return sql<FindingGroup[]>`
    WITH latest_run AS (
      SELECT id FROM atlas.atlas_validator_runs
      WHERE scope = 'frontend_diff' AND status = 'success'
      ORDER BY started_at DESC
      LIMIT 1
    )
    SELECT
      f.route,
      f.surface,
      f.severity,
      COUNT(*)                                      AS count,
      MIN(f.identifier)                             AS sample_identifier,
      MIN(f.expected_value)                         AS sample_expected,
      MIN(f.actual_value)                           AS sample_actual,
      MAX(f.last_seen)                              AS last_seen
    FROM atlas.atlas_validator_findings f
    JOIN latest_run lr ON lr.id = f.run_id
    WHERE f.finding_class IN ('frontend_diff', 'frontend_extract_error')
    GROUP BY f.route, f.surface, f.severity
    ORDER BY
      CASE f.severity WHEN 'P0' THEN 0 WHEN 'P1' THEN 1 WHEN 'P2' THEN 2 ELSE 3 END,
      f.route
  `
}

// Per-route P0/P1/P2 breakdown for the latest successful run.
export async function getLatestRouteSummary(): Promise<RouteSummary[]> {
  return sql<RouteSummary[]>`
    WITH latest_run AS (
      SELECT id FROM atlas.atlas_validator_runs
      WHERE scope = 'frontend_diff' AND status = 'success'
      ORDER BY started_at DESC
      LIMIT 1
    )
    SELECT
      COALESCE(f.route, '(unknown)')                AS route,
      COUNT(*) FILTER (WHERE f.severity = 'P0')     AS p0,
      COUNT(*) FILTER (WHERE f.severity = 'P1')     AS p1,
      COUNT(*) FILTER (WHERE f.severity = 'P2')     AS p2,
      COUNT(*)                                      AS total
    FROM atlas.atlas_validator_findings f
    JOIN latest_run lr ON lr.id = f.run_id
    WHERE f.finding_class IN ('frontend_diff', 'frontend_extract_error')
    GROUP BY f.route
    ORDER BY p0 DESC, p1 DESC, f.route
  `
}

// 7-day trend: P0/P1/P2 counts per day for frontend_diff runs.
export async function getValidatorTrend(days = 7): Promise<DayTrend[]> {
  return sql<DayTrend[]>`
    SELECT
      started_at::date::text                            AS run_date,
      COALESCE(SUM(p0_count), 0)                        AS p0,
      COALESCE(SUM(p1_count), 0)                        AS p1,
      COALESCE(SUM(p2_count), 0)                        AS p2,
      COALESCE(SUM(p0_count + p1_count + p2_count), 0)  AS total
    FROM (
      SELECT
        r.started_at,
        COUNT(*) FILTER (WHERE f.severity = 'P0') AS p0_count,
        COUNT(*) FILTER (WHERE f.severity = 'P1') AS p1_count,
        COUNT(*) FILTER (WHERE f.severity = 'P2') AS p2_count
      FROM atlas.atlas_validator_runs r
      LEFT JOIN atlas.atlas_validator_findings f
        ON f.run_id = r.id
        AND f.finding_class IN ('frontend_diff', 'frontend_extract_error')
      WHERE r.scope = 'frontend_diff'
        AND r.started_at >= NOW() - INTERVAL '1 day' * ${days}
      GROUP BY r.id, r.started_at
    ) sub
    GROUP BY started_at::date
    ORDER BY run_date DESC
  `
}

File: atlas/api/admin/validator.py
"""Phase C Validator API — run listing and finding queries.

Endpoints:
    GET /api/admin/validator/runs
        List recent validator runs (all scopes), most-recent first.
        Query params: limit (default 20, max 100), scope (optional filter).

    GET /api/admin/validator/findings
        Query validator findings with filtering.
        Query params: run_id, severity, finding_class, surface, limit.

Admin-only: uses the same ``_require_admin`` guard as proposals.py.
"""

from __future__ import annotations

import os
from typing import Any

from fastapi import APIRouter, HTTPException, Query, Request
from sqlalchemy import text

router = APIRouter(prefix="/api/admin/validator", tags=["admin"])


def _require_admin(request: Request) -> str:
    """Authorize via JWT or internal secret. Returns reviewer label."""
    user = getattr(request.state, "user", None)
    if user is not None and getattr(user, "role", "") in ("admin", "service_role"):
        return getattr(user, "user_id", "unknown")

    auth_header = request.headers.get("Authorization", "")
    if auth_header.startswith("Bearer "):
        token = auth_header[7:]
        secret = os.environ.get("ATLAS_INTERNAL_SECRET", "")
        if secret and token == secret:
            return "internal-proxy"

    if user is None:
        raise HTTPException(status_code=401, detail="auth required")
    raise HTTPException(status_code=403, detail="admin role required")


@router.get("/runs")
async def list_runs(
    request: Request,
    limit: int = Query(default=20, ge=1, le=100),
    scope: str | None = Query(default=None),
) -> dict[str, Any]:
    """List recent validator runs.

    Args:
        limit: Max runs to return.
        scope: Filter by scope (sensibility, schema_coverage, frontend_diff).
    """
    _require_admin(request)

    from atlas.db import get_engine

    engine = get_engine()
    scope_clause = "AND scope = :scope" if scope else ""

    with engine.connect() as conn:
        rows = conn.execute(
            text(f"""
                SELECT id, started_at, completed_at, status, scope, n_findings
                FROM atlas.atlas_validator_runs
                WHERE 1=1 {scope_clause}
                ORDER BY started_at DESC
                LIMIT :limit
            """),
            {"scope": scope, "limit": limit} if scope else {"limit": limit},
        ).fetchall()

    return {
        "data": [
            {
                "id": str(r[0]),
                "started_at": r[1].isoformat() if r[1] else None,
                "completed_at": r[2].isoformat() if r[2] else None,
                "status": r[3],
                "scope": r[4],
                "n_findings": r[5],
            }
            for r in rows
        ],
        "meta": {"count": len(rows)},
    }


@router.get("/findings")
async def list_findings(
    request: Request,
    run_id: str | None = Query(default=None),
    severity: str | None = Query(default=None),
    finding_class: str | None = Query(default=None),
    surface: str | None = Query(default=None),
    limit: int = Query(default=50, ge=1, le=500),
) -> dict[str, Any]:
    """List validator findings with optional filters.

    Args:
        run_id: Filter by run UUID.
        severity: Filter by severity (P0, P1, P2, P3).
        finding_class: Filter by class (frontend_diff, insensible_value, etc.).
        surface: Filter by surface (e.g. 'stock.conviction_score').
        limit: Max findings to return.
    """
    _require_admin(request)

    from atlas.db import get_engine

    engine = get_engine()

    clauses: list[str] = []
    params: dict[str, Any] = {"limit": limit}

    if run_id:
        clauses.append("run_id = :run_id")
        params["run_id"] = run_id
    if severity:
        clauses.append("severity = :severity")
        params["severity"] = severity
    if finding_class:
        clauses.append("finding_class = :finding_class")
        params["finding_class"] = finding_class
    if surface:
        clauses.append("surface ILIKE :surface")
        params["surface"] = f"%{surface}%"

    where = ("WHERE " + " AND ".join(clauses)) if clauses else ""

    with engine.connect() as conn:
        rows = conn.execute(
            text(f"""
                SELECT id, run_id, finding_class, severity, route, surface,
                       identifier, expected_value, actual_value,
                       delta_abs, delta_pct,
                       first_seen, last_seen
                FROM atlas.atlas_validator_findings
                {where}
                ORDER BY severity ASC, last_seen DESC
                LIMIT :limit
            """),
            params,
        ).fetchall()

    return {
        "data": [
            {
                "id": str(r[0]),
                "run_id": str(r[1]),
                "finding_class": r[2],
                "severity": r[3],
                "route": r[4],
                "surface": r[5],
                "identifier": r[6],
                "expected_value": r[7],
                "actual_value": r[8],
                "delta_abs": str(r[9]) if r[9] is not None else None,
                "delta_pct": str(r[10]) if r[10] is not None else None,
                "first_seen": r[11].isoformat() if r[11] else None,
                "last_seen": r[12].isoformat() if r[12] else None,
            }
            for r in rows
        ],
        "meta": {"count": len(rows)},
    }
